misc/python/materialize/data_ingest/executor.py

```diff
@@ -0,0 +1,359 @@
+
+
+def idfn(d: str, ctx: Any) -> str:
+    return d
+
+
+def delivery_report(err: str, msg: Any) -> None:
+    assert err is None, f"Delivery failed for User record {msg.key()}: {err}"
+    # print('User record {} successfully produced to {} [{}] at offset {}'.format(
+    #    msg.key(), msg.topic(), msg.partition(), msg.offset()))
+
+
+class KafkaExecutor(Executor):
+    producer: Producer
+    avro_serializer: AvroSerializer
+    key_avro_serializer: AvroSerializer
+    topic: str
+    table: str
+    fields: List[Field]
+
+    def __init__(
+        self,
+        num: int,
+        conn: pg8000.Connection,
+        ports: Dict[str, int],
+        fields: List[Field],
+    ):
+        self.topic = f"data-ingest-{num}"
+        self.table = f"kafka_table{num}"
+        self.fields = fields
+
+        schema = {
+            "type": "record",
+            "name": "value",
+            "fields": [
+                {
+                    "name": field.name,
+                    "type": str(field.typ.name(Backend.AVRO)).lower(),
+                }
+                for field in fields
+                if not field.is_key
+            ],
+        }
+
+        key_schema = {
+            "type": "record",
+            "name": "key",
+            "fields": [
+                {
+                    "name": field.name,
+                    "type": str(field.typ.name(Backend.AVRO)).lower(),
+                }
+                for field in fields
+                if field.is_key
+            ],
+        }
+
+        kafka_conf = {"bootstrap.servers": f"localhost:{ports['kafka']}"}
+
+        a = AdminClient(kafka_conf)
+        fs = a.create_topics(
+            [NewTopic(self.topic, num_partitions=1, replication_factor=1)]
+        )
+        for topic, f in fs.items():
+            f.result()
+            print(f"Topic {topic} created")
+
+        schema_registry_conf = {"url": f"http://localhost:{ports['schema-registry']}"}
+        registry = SchemaRegistryClient(schema_registry_conf)
+
+        self.avro_serializer = AvroSerializer(registry, json.dumps(schema), idfn)
+
+        self.key_avro_serializer = AvroSerializer(
+            registry, json.dumps(key_schema), idfn
+        )
+
+        registry.register_schema(
+            f"{topic}-value", Schema(json.dumps(schema), schema_type="AVRO")
+        )
+        registry.register_schema(
+            f"{topic}-key", Schema(json.dumps(key_schema), schema_type="AVRO")
+        )
+
+        self.producer = Producer(kafka_conf)
+
+        conn.autocommit = True
+        with conn.cursor() as cur:
+            cur.execute(
+                f"""CREATE SOURCE {self.table}
+                    FROM KAFKA CONNECTION kafka_conn (TOPIC '{self.topic}')
+                    FORMAT AVRO
+                    USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                    ENVELOPE UPSERT"""
+            )
+        conn.autocommit = False
+
+    def run(self, transactions: List[Transaction]) -> None:
```

Comment:
Long, deeply nested method: Can we extract methods? (optional)

Reply:
There is no other place where the code is used at the moment, so I think that would hinder readability.